### ISSUE TITLE
GDB: Line numbers off, Thread names missing characters, Quarkus: galderz/quarkus/tree/t_sourcecache_container_12283

### ISSUE THREAD

**Issue by Karm** [graalvm/mandrel#141]
## Description
 * After some stepping, line numbers are off. It looks like the line numbers are off as soon as JDK sources are entered.
 * Thread names are missing first characters, e.g. `Thread 4 "ecutor-thread-1"` instead of expected `Thread 4 "Executor-thread-1"`

The former is an issue. The latter is just a nuisance.

## How To Reproduce

Mind the location `/tmp/` or adjust the `set directories`.

`unzip` [quarkus-full-microprofile.zip](https://github.com/graalvm/mandrel/files/5302447/quarkus-full-microprofile.zip) `-d /tmp `
```
cd /tmp/quarkus-full-microprofile
export JAVA_HOME=~/workspaceRH/mandrel-release/mandrel-java11-20.1.0.2.Final/;export GRAALVM_HOME=${JAVA_HOME};export PATH=${JAVA_HOME}/bin:${PATH} 
mvn clean compile package -Pnative -Dquarkus.native.debug.enabled=true
mvn dependency:sources
gdb ./target/quarkus-runner
set directories  /tmp/quarkus-full-microprofile/target/quarkus-native-image-source-jar/sources/:/tmp/quarkus-full-microprofile/target/quarkus-native-image-source-jar/sources/src/
(gdb)  b ConfigTestController.java:33
Breakpoint 1 at 0x40960a: file com/example/quarkus/config/ConfigTestController.java, line 33.
(gdb) run&
shell curl http://localhost:8080/data/config/lookup &

Thread 4 "ecutor-thread-1" hit Breakpoint 1, com.example.quarkus.config.ConfigTestController::getLookupConfigValue()void () at com/example/quarkus/config/ConfigTestController.java:33
33          String value = config.getValue("value", String.class);
s
io.smallrye.config.SmallRyeConfig::getValue(java.lang.String, java.lang.Class)()void () at io/smallrye/config/SmallRyeConfig.java:140
140         return getValue(name, getConverter(aClass));
(gdb) s
io.smallrye.config.SmallRyeConfig::getConverter(java.lang.Class)()void () at io/smallrye/config/SmallRyeConfig.java:272
272         final Converter<?> exactConverter = converters.get(asType);
(gdb) s
java.util.concurrent.ConcurrentHashMap::get(java.lang.Object)()void () at java.base/java/util/concurrent/ConcurrentHashMap.java:936
936         int h = spread(key.hashCode());
(gdb) s
java.lang.Object::hashCode()void () at java.base/java/lang/Object.java:105
105      * @see     java.lang.Object#equals(java.lang.Object)
(gdb) s
java.lang.System::identityHashCode(java.lang.Object)()void () at java.base/java/lang/System.java:297
297      *

(gdb) c&
Continuing.
(gdb) Config value from ConfigProvider lookup value 
```
(the last sentence ending in "lookup value" is expected output of the web app)

e.g. `java.base/java/lang/System.java:297` is actually inside a comment...
```
 294 
 295     /**
 296      * Sets the System security.
 297      *
 298      * If there is a security manager already installed, this method first
 299      * calls the security manager's {@code checkPermission} method
 300      * with a {@code RuntimePermission("setSecurityManager")}
 301      * permission to ensure it's ok to replace the existing
 302      * security manager.
 303      * This may result in throwing a {@code SecurityException}.
 304      *
 305      * <p> Otherwise, the argument is established as the current

```

### Configuration:
 - OS: Fedora
 - GDB: 7.12
 - Architecture: AMD64
 - Mandrel version: 20.1.0.2.Final
 - JDK version: OpenJDK 11.0.8+10

## Expected behavior
Line numbers match the code all the way.

**Comment by Karm** [comment]
The used https://github.com/quarkusio/quarkus/pull/12294 is now in master and 1.7.4.

**Comment by Karm** [comment]
@jerboaa WDYT?

**Comment by jerboaa** [comment]
@Karm I'll take a look.

**Comment by jerboaa** [comment]
@Karm I can confirm the issue. As to what's wrong, I've got no idea :-(

```
(gdb) step
java.lang.Object::hashCode(void) () at java.base/java/lang/Object.java:105
(gdb) list
100	     * distinct objects. (The hashCode may or may not be implemented
101	     * as some function of an object's memory address at some point
102	     * in time.)
103	     *
104	     * @return  a hash code value for this object.
105	     * @see     java.lang.Object#equals(java.lang.Object)
106	     * @see     java.lang.System#identityHashCode
107	     */
108	    @HotSpotIntrinsicCandidate
109	    public native int hashCode();
(gdb) step
java.lang.System::identityHashCode(java.lang.Object)(void) () at java.base/java/lang/System.java:297
297	     *
(gdb) list
292	    private static native void setOut0(PrintStream out);
293	    private static native void setErr0(PrintStream err);
294	
295	    /**
296	     * Sets the System security.
297	     *
298	     * If there is a security manager already installed, this method first
299	     * calls the security manager's {@code checkPermission} method
300	     * with a {@code RuntimePermission("setSecurityManager")}
301	     * permission to ensure it's ok to replace the existing
(gdb) step
298	     * If there is a security manager already installed, this method first
(gdb) list
293	    private static native void setErr0(PrintStream err);
294	
295	    /**
296	     * Sets the System security.
297	     *
298	     * If there is a security manager already installed, this method first
299	     * calls the security manager's {@code checkPermission} method
300	     * with a {@code RuntimePermission("setSecurityManager")}
301	     * permission to ensure it's ok to replace the existing
302	     * security manager.
(gdb)
```

**Comment by jerboaa** [comment]
@adinn FYI. This might be something for you to look at. I'll try to reproduce with upstream master tomorrow.

**Comment by zakkak** [comment]
Regarding thread names, this is expected. Linux kernel does not support more than 15 chars (plus the null terminating char) to be used as thread names. Size defined [here](https://github.com/torvalds/linux/blob/02de58b24d2e1b2cf947d57205bd2221d897193c/include/linux/sched.h#L212)

**Comment by Karm** [comment]
@jerboaa THX .

@zakkak ...and if it was cut and the end, it would be all "Executor-worker" without "-1" or "-23" :)  O.K. Thanks for the explanation. I did not know about the limitation. Non issue then.

**Comment by zakkak** [comment]
Regarding the first issue, it appears that you are hitting a substituted method and the line number makes total sense in the file where the substitution is happening.

`java.base/java/lang/System.java:297` should be `com/oracle/svm/core/jdk/JavaLangSubstitutions.java:297`
https://github.com/graalvm/mandrel/blob/646a5e5a344751321c0c2b809b2bd8c353adb16e/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaLangSubstitutions.java#L297

and
`java.base/java/lang/Object.java:105` should be `com/oracle/svm/core/jdk/JavaLangSubstitutions.java:105`
https://github.com/graalvm/mandrel/blob/646a5e5a344751321c0c2b809b2bd8c353adb16e/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaLangSubstitutions.java#L105

**Comment by zakkak** [comment]
@adinn I don't see any special handling for this so I will work on fixing it. If you are already taking care of this somehow please let me know.

**Comment by jerboaa** [comment]
> Regarding the first issue, it appears that you are hitting a substituted method and the line number makes total sense in the file where the substitution is happening.
> 
> `java.base/java/lang/System.java:297` should be `com/oracle/svm/core/jdk/JavaLangSubstitutions.java:297`
> 
> https://github.com/graalvm/mandrel/blob/646a5e5a344751321c0c2b809b2bd8c353adb16e/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaLangSubstitutions.java#L297
> 
> and
> `java.base/java/lang/Object.java:105` should be `com/oracle/svm/core/jdk/JavaLangSubstitutions.java:105`
> 
> https://github.com/graalvm/mandrel/blob/646a5e5a344751321c0c2b809b2bd8c353adb16e/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaLangSubstitutions.java#L105

Ah, right! Thanks @zakkak

**Comment by zakkak** [comment]
This should now be fixed by oracle/graal#3421
@Karm please reopen if not.
